### ISSUE TITLE
[Bug]: Cannot add hash to redirect url

### ISSUE THREAD

**Issue by andrelandgraf** [remix-run/remix#1641]
### Which Remix packages are impacted?

- `remix` (Remix core)

### What version of Remix are you using?

1.1.3

### What version of Node are you using? Minimum supported version is 14.

16.13.0

### Steps to Reproduce

Add a hash to your loader function or action function redirect. It never reaches the url bar.

```typescript
export const loader = () => {
   return redirect('/#hash');
}
```

### Expected Behavior

The hash that is added to the redirect url should take effect in the browser and be accessible via the `useLocation` hook.

### Actual Behavior

The hash never makes it to the browser and cannot be accessed via the `useLocation` hook.

### Note

SearchParams work just fine. It's just the hash that is not working.

**Comment by andrelandgraf** [comment]
Loaders and Actions seems to have an issue with hash in general. Probably connected to #1574

**Comment by Tapudp** [comment]
Is this open to contribute to?

**Comment by b0b4456** [comment]
related to https://github.com/remix-run/remix/issues/1862

**Comment by machour** [comment]
The hash value is never sent to the server. It's a pure client-side thing.
So you can't use them from loader or action, that are run server-side.

**Comment by a3diti** [comment]
> The hash value is never sent to the server. It's a pure client-side thing. So you can't use them from loader or action, that are run server-side.

The loader does send the hash successfully to the client using redirect since its resolved by remix with the `X-Remix-Redirect` header

![image](https://user-images.githubusercontent.com/3646190/161067684-b03f13a9-71da-49fa-b09a-fb77801a462c.png)
My use case is sending a json encoded on the hash as a query 🗡️

**Comment by a3diti** [comment]
The problem lies here:
https://github.com/remix-run/remix/blob/621b392cc6151be51dc30f4e2bba4603085d5da5/packages/remix-react/routes.tsx#L226-L229
It seems after parsing the redirect string, the transition is not taking the hash anywhere, only `pathname` and `search`

**Comment by machour** [comment]
Reopened this issue as my reason for closing was totally out of topic 😅 

@andrelandgraf I just tried on a fresh remix install, it works for me (the # reaches the url bar)
Please provide a reproduce case for investigation (where are you trying to read the hash value?)

**Comment by a3diti** [comment]
@machour @andrelandgraf 
I checked also with fresh install, same thing occurs, to make it clear how it happens:

We have two routes:

**index.tsx**
```
import { Link } from "remix";
export default function Index() {
  return (
    <div style={{ fontFamily: "system-ui, sans-serif", lineHeight: "1.4" }}>
      <Link to={"/test"}>Lets test</Link>
    </div>
  );
}
```

**test.tsx**
```
import { LoaderFunction, redirect } from "remix";
export const loader: LoaderFunction = () => {
  return redirect("/#hello");
};
```

My actions:
I visit directly `/test` creates a `Location: /#hello` header and redirects me at `/#hello`
I visit `/` and click on `Lets test` link, it sends me to `/test` and then creates location header `X-Remix-Redirect: /#hello` which redirects me to `/` using the transition of remix. It should redirect me to `/#hello`

As I stated before the bug I identified is here, which makes sense because the transition is ignoring the hash from the parsed `X-Remix-Redirect` header
https://github.com/remix-run/remix/blob/621b392cc6151be51dc30f4e2bba4603085d5da5/packages/remix-react/routes.tsx#L226-L229

**Comment by machour** [comment]
My investigation shows that if you change to import from "@remix-run/node", the problem goes away

**Comment by a3diti** [comment]
@machour 
Even after making the changes of https://github.com/remix-run/remix/pull/2581 the transition is not triggering the `hashchange` event

My hashchange event listener is run only once with `useEffect` on a permanently mounted component
For example from my sample before, this is the new **index.tsx**
```ts
import { Link } from "remix";
import { useEffect } from "react";

export default function Index() {
  useEffect(() => {
    window.addEventListener("hashchange", (e) => {
      console.log("hash changed...", e);
    });
  }, []);
  return (
    <div style={{ fontFamily: "system-ui, sans-serif", lineHeight: "1.4" }}>
      <Link to={"/test"}>Lets test</Link>
    </div>
  );
}
```

What is happening is that the URL hash did change from the new patch when transition was triggered by `X-Remix-Redirect`, but the `hashchange` event did not trigger at all.

**Comment by machour** [comment]
@a3diti Are you sure that hashchange is triggered when the whole url changes ? (you did go through a redirect, and you got back to a new url with a hash baked in)

**Comment by a3diti** [comment]
> @a3diti Are you sure that `hashchange` is triggered when the whole url changes ? (you did go through a redirect, and you got back to a new url with a hash baked in)

@machour 
The `hashchange` event is **NOT** triggered, which raises a new issue to the redirect transition with hash

For me the temporary solution was to make the patch with [url.hash](https://github.com/remix-run/remix/pull/2581/commits/7807af4322b2758e4e370b18c4f2de4435977958#diff-d10f27b26c030cb47343fa74a70c39accd9e2dccc6cdd8df40c0f3bee230ca61) on transition and on the client instead of using a `window.addEventListener("hashchange", hashchange)`, I switched to a interval that checks for a new hash change very 1sec.

**Comment by machour** [comment]
yeah, I meant are you sure it **SHOULD** be triggered?

- From my understanding, `hashchange` is triggered when you're **already** on the page, and the hash changes.

```html
<!-- foo.html -->
<script>
 window.addEventListener("hashchange", (e) => {
      console.log("hash changed...", e);
    });
</script>
```

1. If I load `foo.html#bar` in the browser, I see no console log.
2. If I manually change it, I see the log

In your case, you are being redirected to a new page with a hash, so you're in case 1, and `hashchange` is not triggered.

**Comment by machour** [comment]
Closing this as the original reported issue have been fixed.
@a3diti Feel free to open a new issue if you're sure this is a remix issue, and thank you for the pointers!
